samples/test-dal-with-nunit/DalWithNunitSample/DALTests/PizzaServiceTests.cs

```diff
@@ -0,0 +1,103 @@
+﻿using System.Threading.Tasks;
+using DAL;
+using FluentAssertions;
+using Microsoft.EntityFrameworkCore;
+using NUnit.Framework;
+
+namespace DALTests
+{
+    public class PizzaServiceTests : TestBase
+    {
+        private IPizzaService _pizzaService;
+
+        [OneTimeSetUp]
+        public void Initialization()
+        {
+            _pizzaService = new PizzaService(DbContext);
+        }
+
+        [Test]
+        public async Task GetAll_WhenPizzasListExists_ShouldReturnAllPizzas()
+        {
+            // Arrange
+            DbContext.Pizzas.AddRange(new Pizza[]
```

Comment:
Such arrange/assert is not thread safe as you might query the data seeded by other tests. They just might be run in parallel.

Reply:
As far as i know by default NUnit executes tests one by one, not in parallel. Let's leave it simple

Reply:
I disagree.  The test interface itself is async.  Any reasonable person would want to run their tests synchronous.  This whole PR is a mess of choices and mistakes and shortcuts.  Just do it right and everyone benefits long term.

Reply:
The right way to do this is to EnsureTransaction around each DbContext transaction, and roll it back after you are done, IMHO.

Reply:
Sorry that last comment came off harsh, by the way.  I just got frustrated when I gave feedback and rather than implement the feedback, something completely worse in my eyes was done instead.  Now we have no DAL abstraction in the test, just entity framework directly used in BLL.  If EFSuperCore comes out in 5 years, on a very large code base, this will be nightmare to refactor.  This is why Microsoft can never upgrade anything easily, because customers take shortcuts and to fix their shortcuts is not economically possible because no developer can tell you how long it will take to refactor services to support new DAL.  Repository is supposed to do that.  You just count how many repositories you have and do straight math on how long to upgrade.  I've seen this problem a thousand times.

Reply:
So, you are proposing to implement classic generic repository over EfCore and use it inside our service?

Reply:
It could be a generic repository, or just a repository with an IRepository marker interface.

A good example app is ASP.NET Boilerplate project on GitHub.

It might not be a bad strategy to promote your guys awesome work here (I love this project so much) to submit PRs to projects like ASP.NET Boilerplate demonstrating TestEnvironment.Docker.  https://aspnetboilerplate.com/Samples https://github.com/aspnetboilerplate/aspnetboilerplate